FirebaseTestProject/Models/Post.swift

```diff
@@ -49,13 +79,37 @@ struct PostService {
         return db.collection("posts")
     }
     
+    static var favoritesReference: CollectionReference {
+        let db = Firestore.firestore()
+        return db.collection("favorites")
+    }
+    
     static func getPosts() async throws -> [Post] {
         let postsSnapshots = try await postsReference.getDocuments()
         let posts = postsSnapshots.documents.map { Post(from: $0.data()) }
         return posts
     }
     
+    static func getFavorites() async throws -> [Post] {
+        let userid = UserDefaults.standard.value(forKey: "userid") != nil ?  UserDefaults.standard.value(forKey: "userid") as! String : "00854E9E-8468-421D-8AA2-605D8E6C61D9"
+        let favoritesQuery = favoritesReference.whereField("userid", isEqualTo: userid)
+        let favoritesSnapshots = try await favoritesQuery.getDocuments()
+        let favorites = favoritesSnapshots.documents.map { Post(from: $0.data()) }
+        
+        return favorites
+    }
+    
     static func upload(_ post: Post) async throws {
         try await postsReference.document(post.id.uuidString).setData(post.jsonDict)
     }
+    
+    static func favorite(_ post: Post) async throws {
+        let userid = UserDefaults.standard.value(forKey: "userid") as? String ?? "00854E9E-8468-421D-8AA2-605D8E6C61D9"
+        try await favoritesReference.document(post.id.uuidString).setData(Favorite(postid: post.id.uuidString, userid: userid).jsonDict)
```

Comment:
Please see the note above regarding using the post ID as a primary key for favorites.

Reply:
I considered the same idea. But, theoretically, since a user can only favorite a post at most once, the combination of userid+postid intuitively creates a unique value in itself. For example, if we try to reference Firebase for a favorite posts, we will filter the Favorites collection query to only include where userid = currentuser, all returned posts will be a unique Post.

Reply:
LMK: It would be easier to follow and understand with a separate id but I just considered this route and thought it to be less clunky.

Reply:
Regarding primary keys: it appears that only one `Favorite` can be saved to Firebase (because the record is referenced as `favoritesReference.document(post.id.uuidString)`).

If you don’t want to add a separate `id` property, I would just make sure that the `favorite(_:)` and `unfavorite(_:)` methods on the `PostService` are referencing the correct record. For example, you could:
* pass both the user and post IDs to `favoritesReference.document(_:)`
* use other methods in place of `favoritesReference.document(_:)`- for `favorite(_:)`, you could use `favoritesReference.addDocument(data:)`; for `unfavorite(_:)`, you could use a query to match both the post ID and the user ID